src/buffer.rs

```diff
@@ -0,0 +1,294 @@
+///
+/// # Allocations
+///
+/// By default, this buffer expands every time one byte is written. The `with_increment`
+/// constructor can be used to change how much the buffer expands to reduce overall number of allocations.
+///
+/// # Pwrite and TryIntoCtx
+///
+/// Types must implement `TryIntoCtx` with a second generic parameter of `DynamicBuffer` to be used.
+///
+/// ```rust
+/// use scroll::{ctx::TryIntoCtx, DynamicBuffer, Error, LE, Pwrite};
+///
+/// struct EvenOdd(u32);
+///
+/// impl TryIntoCtx<(), DynamicBuffer> for EvenOdd {
+///     type Error = Error;
+///
+///     fn try_into_ctx(self, buf: &mut DynamicBuffer, _: ()) -> Result<usize, Self::Error> {
+///         let offset = &mut 0;
+///
+///         if self.0 % 2 == 0 {
+///             buf.gwrite_with(self.0, offset, LE)?;
+///         } else {
+///             buf.gwrite_with(self.0 as u64, offset, LE)?;
+///         }
+///
+///         Ok(*offset)
+///     }
+/// }
+///
+/// let mut buf = DynamicBuffer::new();
+///
+/// let offset = buf.pwrite(EvenOdd(2), 0).unwrap();
+/// assert_eq!(buf.get(), [2, 0, 0, 0]);
+/// buf.pwrite(EvenOdd(3), offset).unwrap();
+/// assert_eq!(buf.get()[offset..], [3, 0, 0, 0, 0, 0, 0, 0]);
+/// ```
+///
+/// ## Why can't I use types that already implement the default TryIntoCtx?
+///
+/// The default type parameters for `TryIntoCtx` expose a mutable raw slice to write into.
+/// This is great for pre-allocated buffers and types that have a static written size; however,
+/// we cannot create a custom slice type that performs dynamic expansions under the hood.
+///
+/// As a result, we need to expose a separate writing type that can track offsets and calculate when to allocate.
+///
+/// However, if your `TryIntoCtx` impls don't use any special slice APIs and just use `Pwrite` and/or
+/// basic indexing, it's extremely easy to migrate! Just add the `DynamicBuffer` generic type.
+pub struct DynamicBuffer {
+    buffer: Vec<u8>,
+    alloc_increment: usize,
+    start_offset: usize,
+    write_end: usize,
+}
+
+impl DynamicBuffer {
+    /// Constructs an empty `DynamicBuffer` with the default allocation increment of one byte.
+    pub fn new() -> DynamicBuffer {
+        Self::with_increment(1)
+    }
+
+    /// Constructs an empty `DynamicBuffer` with a custom allocation increment.
+    pub fn with_increment(alloc_increment: usize) -> DynamicBuffer {
+        DynamicBuffer {
+            buffer: vec![],
+            alloc_increment,
+            start_offset: 0,
+            write_end: 0,
+        }
+    }
+
+    /// Gets a slice of all written bytes. Does not include extra bytes allocated by a large increment.
+    /// ```rust
+    /// use scroll::{DynamicBuffer, LE, Pwrite};
+    /// let mut buf = DynamicBuffer::new();
+    ///
+    /// buf.pwrite_with(2u32, 0, LE).unwrap();
+    /// assert_eq!(buf.get(), [2, 0, 0, 0]);
+    /// ```
+    pub fn get(&self) -> &[u8] {
+        &self.buffer[..=self.write_end]
+    }
+}
```

Comment:
I'm on fence about adding:
```rust
pub fn into_inner(self) -> Vec<u8> {
  self.buffer
}
```

on one hand it's nice and typical api of this sort; on other hand it forever forces the implementation to either use a Vec<u8> internally or create one on the fly if the user calls this method. It's probably fine?